pinot-broker/src/main/java/org/apache/pinot/broker/queryquota/MaxHitRateTracker.java

```diff
@@ -0,0 +1,80 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.broker.queryquota;
+
+import com.google.common.annotations.VisibleForTesting;
+
+/**
+ * A stateful version of hit counter. Similar to the default hit counter, it maintains a list of buckets.
+ * Whereas it maintains an extra variable called _lastAccessTimestamp which tracks the last access time.
+ * If the stateful hit counter gets queried, it firstly compares the current timestamp and the last access timestamp,
+ * calculating the start index and end index among the buckets. Then, it traverses through all the valid candidate buckets.
+ * If the current timestamp has exceeded the current time range of all the buckets, this hit counter will use
+ * the current timestamp minus the default time queried time range to calculate the start time index.
+ */
+public class MaxHitRateTracker extends HitCounter {
+  private static int ONE_SECOND_BUCKET_WIDTH_MS = 1000;
+  private static int MAX_TIME_RANGE_FACTOR = 2;
+
+  private final long _maxTimeRangeMs;
+  private final long _defaultTimeRangeMs;
+  private volatile long _lastAccessTimestamp;
+
+  public MaxHitRateTracker(int timeRangeInSeconds) {
+    this(timeRangeInSeconds, timeRangeInSeconds * MAX_TIME_RANGE_FACTOR);
+  }
+
+  public MaxHitRateTracker(int defaultTimeRangeInSeconds, int maxTimeRangeInSeconds) {
+    super(maxTimeRangeInSeconds, (int) (maxTimeRangeInSeconds * 1000L / ONE_SECOND_BUCKET_WIDTH_MS));
+    _defaultTimeRangeMs = defaultTimeRangeInSeconds * 1000L;
+    _maxTimeRangeMs = maxTimeRangeInSeconds * 1000L;
+  }
+
+  /**
+   * Get the maximum count among the buckets
+   */
+  public int getMaxCountPerBucket() {
+    return getMaxCountPerBucket(System.currentTimeMillis());
+  }
+
+  @VisibleForTesting
+  int getMaxCountPerBucket(long now) {
```

Comment:
should this method be synchronized?
I would code the method like this:
- Get the value of _lastAccessTimeStamp in the beginning of the method (`then = _lastAccessTimeStamp`)
- Use `then` throughout the method
- Set the` _lastAccessTimestamp` to `now` at the end of the method

That will protect us against multiple calls, if any (just in case).

it will also prevent us from accessing the volatile variable repeatedly.

Reply:
No, it doesn't need to be synchronized.
The callback function will be called by 1 single thread. Thus, `_lastAccessTimeStamp` will also be modified by the same thread. 
The bucket belonging to the end index won't be queried, so there is no need to add the block on it. Plus, all the buckets have already been in AtomicIntegerArray. There is no need to add extra protection